cohortextractor/query_engines/mssql.py

```diff
@@ -404,6 +412,49 @@ def get_case_expression(self, mapping, default):
             else_=default,
         )
 
+    def get_expression_for_value_from_function(self, value):
+        argument_expressions = []
+        tables = set()
+        for arg in value.arguments:
+            arg_expr, arg_tables = self.get_value_expression(arg)
+            argument_expressions.append(arg_expr)
+            tables.update(arg_tables)
+
+        # TODO: I'd quite like to build this map by decorating the methods e.g.
+        #
+        #   @handler_for(DateDifferenceInYears)
+        #   def my_handle_fun(...)
+        #
+        # but the simple thing will do for now.
+        class_method_map = {DateDifferenceInYears: self.date_difference_in_years}
+
+        try:
+            method = class_method_map[value.__class__]
+        except KeyError:
+            raise ValueError(f"Unsupported function: {value}")
+        value_expression = method(*argument_expressions)
+
+        # TODO: Once we have consistent hashing for QueryNodes we can remove
+        # this sort. This is not required for correctness, just for consistent
+        # test recordings.
+        tables = sorted(tables, key=lambda i: i.name)
```

Comment:
Are the tables from a `ValuefromFunction`'s arguments guaranteed to always be `BaseTable`s, with a `name`, and not a `FilteredTable`?

Reply:
I guess they are for now, since we only have the `age_as_of` method that uses this, and it can only be called on the patient table

Reply:
This is a good catch, but not quite for the reason you think! In fact, my comment is totally wrong because these are instances of `sqlalchemy.Table` rather than `query_language.Table`. So they are guaranteed to have a `name` attribute, but fixing the hashing on QueryNode won't make any difference here. A good illustration of how confusing this codebase can be!